### ISSUE TITLE
React 18: bootstrapScriptContent escapes HTML so quotes can’t be used

### ISSUE THREAD

**Issue by calebmer** [facebook/react#23063]
If you use the `bootstrapScriptContent` option of `renderToPipeableStream()` to provide hydration data (as [recommended in the <script> upgrade guide](https://github.com/reactwg/react-18/discussions/114)) with `JSON.stringify()` it doesn’t work because React escapes HTML characters in `bootstrapScriptContent` like quotes. I’ve worked around this by using backticks to deliniate strings.

What’s the correct thing to do here?

https://github.com/facebook/react/blob/cdb8a1d19d0c0d43a72c3f0fe739b04da247c360/packages/react-dom/src/server/ReactDOMServerFormatConfig.js#L99-L105

**Comment by salazarm** [comment]
@sebmarkbage @acdlite Is the `escapeTextForBrowser` usage [here](https://github.com/facebook/react/blob/main/packages/react-dom/src/server/ReactDOMServerFormatConfig.js#L104) intentional? I think it might make sense not to escape here since it's not `HTML` and its supposed to be javascript here.

**Comment by sebmarkbage** [comment]
Yea it's probably not the correct escape. We should escape somehow though because `</script>` and such content shouldn't be allowed. Not escaping random JSON that might contain user generated strings would be a bad and common security issue.

It's a common security issue in many non-meta frameworks.

cc @sophiebits knows the exact pattern needed.

If we can't make it generic we should at least rename it "dangerous" and describe how to escape any JSON content someone might add there.

**Comment by sophiebits** [comment]
See the update at the end of my post here: https://sophiebits.com/2012/08/03/preventing-xss-json.html. Basically replacing the s in `</script` with \u0073 (or S with \u0053).

It's security-safe (to my belief) but note that this technically can change the meaning of a script with a tagged template literal since the raw text is available, but I doubt that would come up in practice so it's probably fine.

The only other ideas I have are having the script encoded in some fully out-of-band way like a base64 data URI or putting the whole script in a string and then eval-ing it (both of which are weird for CSP).

**Comment by sebmarkbage** [comment]
@devknoll @timneutkens What does Next.js do to escape JSON? Presumably that has received enough exposure.

**Comment by sophiebits** [comment]
For JSON you only need to escape / as \/ (many JSON encoders do this automatically).

**Comment by timneutkens** [comment]
This is what we have in Next.js to escape JSON:
https://github.com/vercel/next.js/blob/5f4947e5a2c75099f7466a6ba8274d8ab4240e6c/packages/next/server/htmlescape.ts#L14-L16
https://github.com/vercel/next.js/blob/canary/packages/next/pages/_document.tsx#L817

**Comment by tranvansang** [comment]
This is blocking me from the upgrade to react 18.

How about accept an object with `__htlm` key?

```javascript
{
  bootstrapScriptContent: {
    __dangerouslySetInnerHTML: 'window.__BOOTSTRAP_REACT_APP__("complex inline ssr data")`
  }
}
```

**Comment by sebmarkbage** [comment]
@sophiebits I found this little tidbit while reading the spec. https://github.com/w3c/html/issues/1617

I'm not sure you can use that to inject new scripts but you can certainly cover up future scripts and break the page.

**Comment by sebmarkbage** [comment]
This API only really promises to encode valid JS into HTML. It doesn't promises to turn JSON into JS. In practice anyone using this to inject JSON will need to use another tool to escape JSON into valid JS since we can't parse out the JSON parts from arbitrary JS anyway.

So it's not super critical that we cover all cases because they'll be covered by that encoding anyway. Our escaping is more of a convenience and extra layer of protection.

Escaping the s in `</script>` and `<script>` to `\u0073` as well as the upper-case version is probably sufficient.

**Comment by sebmarkbage** [comment]
I think we're just going to deprecate XHTML support therefore I don't think `]]>` is a sequence of concern. Correct me if I'm wrong.

Additionally, some escape `<!--` and `-->` in JSON because they [can switch the parsing modes](https://github.com/OWASP/json-sanitizer/pull/20). However, since we escape the entire script and not just the JSON part of it, I don't think that concern applies here.

The reason we wouldn't escape them is because the clever hack doesn't work on these while it remains valid JS.

**Comment by gnoff** [comment]
@calebmer should be available in the next minor

**Comment by gaearon** [comment]
Fixed in 18.1.0.
